File: Level 7 - System Design & Probabilistic DS/cuckoo_filter/core_problems.java
import java.util.*;

/**
 * Core Cuckoo Filter Problems in Java
 * Focuses on Empirical False Positive Rate (FPR) measurement.
 */
public class core_problems {

    public static void main(String[] args) {
        int capacity = 1000;
        basic_operations filter = new basic_operations(capacity);

        Set<String> inserted = new HashSet<>();

        // 1. Fill to 50% capacity
        System.out.println("Inserting 500 items...");
        for (int i = 0; i < 500; i++) {
            String item = "item_" + i;
            if (filter.insert(item)) {
                inserted.add(item);
            }
        }

        // 2. Measure False Positives
        int falsePositives = 0;
        int trials = 10000;
        System.out.println("Running " + trials + " trials for non-existent items...");
        for (int i = 0; i < trials; i++) {
            String testItem = "ghost_" + i;
            if (filter.lookup(testItem)) {
                falsePositives++;
            }
        }

        double fpr = (double) falsePositives / trials * 100;
        System.out.println("False Positives: " + falsePositives);
        System.out.println("Empirical FPR: " + String.format("%.4f", fpr) + "%");
        System.out.println("Theoretical FPR (8-bit fp): ~0.78%");

        // 3. Test Deletion Impact
        System.out.println("\nDeleting 100 items...");
        for (int i = 0; i < 100; i++) {
            filter.delete("item_" + i);
        }

        System.out.println("Lookup of deleted item_0: " + filter.lookup("item_0"));
        System.out.println("Lookup of existing item_200: " + filter.lookup("item_200"));
    }
}

File: Level 7 - System Design & Probabilistic DS/cuckoo_filter/core_problems.cpp
#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <iomanip>

/**
 * Core Cuckoo Filter Problems in C++
 * Focuses on False Positive Rate measurement.
 */

// We reuse the logic from basic_operations.cpp (assuming it's available or implemented similarly)
#include "basic_operations.cpp" // Simplified for the example, usually would be a header

int main() {
    int capacity = 1000;
    CuckooFilter cf(capacity);

    std::cout << "Inserting 500 items..." << std::endl;
    for (int i = 0; i < 500; ++i) {
        cf.insert("data_" + std::to_string(i));
    }

    int fp_count = 0;
    int trials = 10000;
    std::cout << "Testing 10,000 random items..." << std::endl;
    for (int i = 0; i < trials; ++i) {
        if (cf.lookup("test_" + std::to_string(i))) {
            fp_count++;
        }
    }

    double fpr = (double)fp_count / trials * 100;
    std::cout << "False Positives: " << fp_count << std::endl;
    std::cout << "Empirical FPR: " << std::fixed << std::setprecision(4) << fpr << "%" << std::endl;

    return 0;
}
